### DISCUSSION TITLE
Thoughts on comprenhension iterator-expressions?

### DISCUSSION THREAD

**Discussion by pedro-gilmora** [dotnet/csharplang#7634] [Polls]
I think that it should be great to have some sugar syntax for simple collections dynamic initializers with the iterator-expression of your choice like

```cs
List<ItemDto> itemsCopy = [..foreach(var item in items) yield item.toDto()];
```

it could be `while` or `for`

**Comment by huoyaoyuan** [comment]
This feels like a combination of collection expression spreads and improved, pre-lowered LINQ.

Currently with LINQ, the followings are allowed:

```csharp
xxx = [..items.Select(item => item.toDto())];
```

or with query syntax:
```csharp
xxx = [..(from item in items select item.toDto())];
```

or method chain syntax:
```csharp
xxx = items.Select(item => item.toDto()).ToXXCollection();
```

I think allowing LINQ to be lowered into more efficient form (#2482) should help more.

**Reply by pedro-gilmora** [reply]
no man, there's no `while` or `for` to express different kind of iterations

**Reply by HaloFour** [reply]
You would have to demonstrate why a query comprehension syntax needs support for those different kinds of iterations, and what use cases that would satisfy that aren't already met via custom iterators.

**Reply by jnm2** [reply]
I'm already writing code like `[.. from item in items select item.toDto()]` today (a direct translation of your example to C# 12 syntax). Having a second set of keywords to do the exact same thing seems like an outcome that is specifically worth avoiding.

`for` would be `from item in Enumerable.Range(...)`, right?

`while` would be the `Enumerable.TakeWhile` method which has no LINQ keyword, but there's a proposal around here somewhere for that IIRC. Proposals to flesh out LINQ syntax with keywords like that would probably be the direction to take. This isn't specific to collection expressions.

**Reply by pedro-gilmora** [reply]
I know, it's just that that solution introduces extra types in compile time

**Reply by jnm2** [reply]
Can you give an example of what "extra types at compile time" would be, and why that would be bad?

**Reply by pedro-gilmora** [reply]
This code 
```cs
List<Student> items = [];
var isOld = false;
List<StudentDto> list = [..from item in items where item.IsOld = isOld && item is { Absent: true } && IsGoodStudent(item) select ToStudentDto(item)];
```
generates:
```cs
[CompilerGenerated]
private sealed class <>c__DisplayClass0_0
{
    public bool isOld;

    public C <>4__this;

    internal bool <M>b__0(Student item)
    {
        int num;
        if (isOld)
        {
            if (item != null && item.Absent)
            {
                num = (<>4__this.IsGoodStudent(item) ? 1 : 0);
                goto IL_0025;
            }
        }
        num = 0;
        goto IL_0025;
        IL_0025:
        bool result = (byte)num != 0;
        item.IsOld = (byte)num != 0;
        return result;
    }

    internal StudentDto <M>b__1(Student item)
    {
        return <>4__this.ToStudentDto(item);
    }
}

<>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
<>c__DisplayClass0_.<>4__this = this;
List<Student> source = new List<Student>();
<>c__DisplayClass0_.isOld = false;
List<StudentDto> list = new List<StudentDto>();
IEnumerator<StudentDto> enumerator = Enumerable.Select(Enumerable.Where(source, new Func<Student, bool>(<>c__DisplayClass0_.<M>b__0)), new Func<Student, StudentDto>(<>c__DisplayClass0_.<M>b__1)).GetEnumerator();
try
{
    while (enumerator.MoveNext())
    {
        StudentDto current = enumerator.Current;
        list.Add(current);
    }
}
finally
{
    if (enumerator != null)
    {
        enumerator.Dispose();
    }
}
```

while mine 
```cs
List<Student> items = [];
var isOld = false;
List<StudentDto> list = [
    ..foreach(var tem in items){
            if(item is { Absent: true } && item.IsOld = isOld && IsGoodStudent(item) 
                yield ToStudentDto(item);
    }
];
```
would just generate
```cs
List<Student> list = new List<Student>();
bool flag = false;
List<StudentDto> list2 = new List<StudentDto>();
List<Student>.Enumerator enumerator = list.GetEnumerator();
try
{
    while (enumerator.MoveNext())
    {
        Student current = enumerator.Current;
        if (current != null && current.Absent && current.IsOld == flag && IsGoodStudent(current))
        {
            list2.Add(ToStudentDto(current));
        }
    }
}
finally
{
    ((IDisposable)enumerator).Dispose();
}
```

**Reply by CyrusNajmabadi** [reply]
We can always adjust generation however we want, as long as the semantics hold. Collection literals have taken a strong stance there. So we sound drive this from the direction of: what do we want the language experience to be.  If we're already there, then no need for lang change, we can just work on the compiler side.

**Reply by jnm2** [reply]
> would just generate

It might go even further and use the `CollectionsMarshal.AsSpan<T>(List<T>)` method and loop over the span, rather than going through the `List<T>.Enumerator` struct. (Depends on an open question about removing version checking from the enumerator)

**Reply by pedro-gilmora** [reply]
That's all about in this poll... enrich language experience, because lambdas and chaining methods seem not so natural for me. The way I see it. 
Oohh.. it's gonna take time and effort to do... so, just measure the acceptance it would have and the impact of the required changes for lowering and optimizations . IDK.. just guessing

**Reply by HaloFour** [reply]
> enrich language experience, because lambdas and chaining methods seem not so natural for me. The way I see it.

You don't have to use lambdas or chained methods.  You can use LINQ, the existing query comprehension syntax.

**Reply by pedro-gilmora** [reply]
I showed you what extra things introduces such approach. I'm not saying it's wrong or too much expensive, it's just that I feel I'm taking a cleaner approach of the existing scope, and maybe optimized. That's my concern. understand it as I read it and take adventage of the same existing scope. For me, feels natural like I said. It's my POV, I was there where you say that. but I know other languages it might sound atractive for other ones to do it in this way.

**Reply by HaloFour** [reply]
> I showed you what extra things introduces such approach. I'm not saying it's wrong or too much expensive

Those are implementation details.

> it's just that I feel I'm taking a cleaner approach of the existing scope, and maybe optimized.

That's your opinion.  It's not enough to justify adding a second flavor of query comprehensions to the language.  That's simply not going to happen.  If you want to suggest specific enhancements to the existing LINQ query comprehensions then maybe those could be taken into advisement, and you'd automatically be able to compose that with collection literal expressions.  However, if your goal is to have the language explicitly support a pattern which is considered non-idiomatic, it's very unlikely to gain any traction, especially since you can already bridge those patterns trivially.

**Reply by HaloFour** [reply]
If C# were to gain inline iterators, it would be orthogonal to collection literal expressions.  It would be expected to be able to assign them directly to a variable, pass them around, use them with LINQ, etc.  They would be expected to behave as normal iterators would, with deferred execution.  As such, I don't see any proposal to add inline iterators within a collection literal to have any different semantics to any other iterator, whether it's declared as a local function or otherwise.

**Reply by pedro-gilmora** [reply]
oh man, that's your sight, I see you don't get the point, while I get yours. but not a problem, you must know the work/effort it takes, but I'm sure it's not bigger than the pattern matching one. So, no big deal. you're not thinking on multiple statements. and I don't think it's orthogonal, in fact you're just building an expression, what else would take to enrich with same natural languaje. Many others langs have list comprenhensions witout having to transform into methods like LINQ does

**Reply by HaloFour** [reply]
Ultimately, collection expressions _already_ work with LINQ.  The language already provides a solution here.  Another solution isn't going to be considered.  Solutions that benefit the whole of LINQ might be, if they can demonstrate that they are worth the effort.

But that's fine, your proposal is out there.

**Reply by pedro-gilmora** [reply]
I see it like a radical position, it would be nicer to mantain natural expressions, rather than potentiating a whole query lib. I see it both related. I know the benefits and the usage of LINQ of course. but I guess you don't think of engagement that would bring. then you'll have to potentiate query syntax to be more loveable and give more clauses and keyworkds, which I find rediculous if you already have the language itself

**Comment by KennethHoff** [comment]
It feels wrong inlining statements like this

**Reply by pedro-gilmora** [reply]
they don't need to be inlined if you want, but should yield the thing we need, that's why I set a poll option saying "can be improved"

**Comment by HaloFour** [comment]
C# already has this syntax, in the form of LINQ.  You can already spread such an expression within a list literal.  I would suggest that it's massively unlikely for the team to consider designing a completely different syntax to accomplish the same thing.

**Reply by pedro-gilmora** [reply]
would it be more performant to have an extra ienumerable method yielding values?, also, I would consider it for small transformation ops

**Reply by HaloFour** [reply]
You'd be much more likely to get traction with specific suggestions to make LINQ more performant in these scenarios than to prescribe a completely new comprehension syntax.

**Reply by pedro-gilmora** [reply]
I know lately LINQ had some improvements, but what I see it's something

Eg:
```cs
List<int> list = [1,2,3,4];
```
gets lowered into
```cs
List<int> list = new List<int>(4);
list.Add(1);
list.Add(2);
list.Add(3);
list.Add(4);
```

Then. something like 
```cs
List<StudentDto> studentsDto= [..while(NextStudent(out var student)) yield ToStudentDto(student)];
```
whould go lower into something like
```cs
List<StudentDto> list = new List<StudentDto>();
Student student;
while (NextStudent(out student))
{
    list.Add(ToStudentDto(student));
}
```

**Reply by HaloFour** [reply]
Collection literals and LINQ are orthogonal concerns.  I would not expect any changes to any comprehension syntax that would take place outside of LINQ.

**Reply by jnm2** [reply]
I'll also say, I'm not sure it's a common use case to do `while (NextThing(out var thing))` to build a collection, but this use case is also something that a helper method would excel at:

```cs
List<StudentDto> list = Helpers.ToEnumerable(NextStudent).Select(ToStudentDto).ToList();
```

**Reply by pedro-gilmora** [reply]
Hey, it's OK to not being agree or expect it. I'm just getting some feedback about this proposal being considered as easy coding

**Reply by NetMage** [reply]
@HaloFour Unfortunately LINQ (especially query syntax) often feels like a sad, neglected corner of C#, though I have been glad to see some Enumerable extensions enhanced in the last few .Net (Core) releases. Without `Expression` improvements, LINQ is stuck in the old C# world.

**Comment by pedro-gilmora** [comment]
> I'll also say, I'm not sure it's a common use case to do `while (NextThing(out var thing))` to build a collection, but this use case is also something that a helper method would excel at:
> 
> ```cs
> List<StudentDto> list = Helpers.ToEnumerable(NextStudent).Select(ToStudentDto).ToList();
> ```

Ok, I see you're not very fan of sugar syntax.. 😅

**Reply by jnm2** [reply]
Language sugar should come into the picture when several factors all combine. The sugar should be for a prevalent pattern, and it should ideally compose really well with existing features rather than only solving exactly one problem. These are just two factors and there are more.

An example of sugar that seems closer to something I'd get behind is a `while` keyword inside LINQ expressions, mapped to `Enumerable.TakeWhile`. That composes really well with existing comprehension expressions in use today, and I've felt the lack on several occasions. However, that is pretty different than what you're talking about, since in my own cases I haven't ever had an `out` parameter. I'm not sure that this stateful conditional-producer-method pattern is common?

But here, also, this is the same thing:
```cs
List<StudentDto> list = [.. from student in Helpers.ToEnumerable(NextStudent) select ToStudentDto(student)];
```

It feels to me like the verbosity is fitting for the style of operations being done: stateful conditional producer method, and a transform method, and buffer into contiguous memory.

Even if the following feature existed, I don't feel like it has significantly changed the verbosity or significantly changed the ease of writing such code:

```cs
List<StudentDto> list = [.. while NextStudent(out var student) select ToStudentDto(student)];
```

**Reply by jnm2** [reply]
Also, extension methods on method groups (and other things) are planned to be investigated for C# 13. With that feature, you'd be able to write:

```cs
List<StudentDto> list = [.. from student in NextStudent.AsEnumerable() select ToStudentDto(student)];
```

**Reply by HaloFour** [reply]
or, if you find the query comprehension syntax a tad verbose for that simple case:

```cs
List<StudentDto> list = [.. NextStudent.AsEnumerable().Select(ToStudentDto)];
```

**Reply by pedro-gilmora** [reply]
you're just not exploding the potencial of comprenhension expressions and how much can reduce overhead at compile time. (I think)
```cs
List<StudentDto> list = [.. while(NextStudent(conditions, out var student) yield ToStudentDto(student, context)];
```

**Reply by HaloFour** [reply]
You're going to have to demonstrate that a completely new comprehension syntax actually provides those kinds of benefits.  And frankly, even if it did, it would be something wrapped up into LINQ, not a completely new and completely superfluous syntactic form.  C# isn't designed that way.

**Reply by pedro-gilmora** [reply]
I'm thinking of something like switch-expressions but thought as collections expressions (like builder pipeline, with multiple iterators and constants). but LINQ lambdas introduces extra types in compile time like I said before while my proposal don't.

**Reply by jnm2** [reply]
Ah, it's the compiler-generated closure classes that you're trying to avoid, using in-place codegen rather than using calls to Enumerable.Where/Select/Range/etc? I think that may have been investigated, but I'm not sure. If the performance benefits can be shown to speak for themselves, then LINQ is the place to do it because it'll improve perf everywhere.

**Comment by En3Tho** [comment]
F# has this for example and it's very useful:

```fsharp
let values = [
    for ... do
        yield ...

    while ... do
       yield ...

    yield 1

    if ... then
        yield 2
        yield 3

   etc ...
]
```

It would be nice to have a way to use something like that in C# too (not sure about syntax though):
```csharp
List<int> values = [
     1, 2, 3,
     ..{ // basically an inlined IEnumeable syntax ?
          for (int i = 0; i < 10; i++) { yield return i; }
          if (...) yield return 1;
          while (...) yield return 2;
     }
]
```

**Reply by pedro-gilmora** [reply]
that's what I'm talking

**Reply by HaloFour** [reply]
C# isn't going to adopt F#'s syntax.

C# has a query comprehension syntax.  It's not going to add a second query comprehension syntax.

**Reply by HaloFour** [reply]
At best you can argue for inline iterators, although the C# team hasn't been that interested in that before.  They consider local functions to be suitable to meet those needs.

**Reply by pedro-gilmora** [reply]
I don't see them comfortable and also not taking a clean approach of the scope it belongs (capturing variables and introduce them in extra classes). Like I said, it reduces the extra type creations and I found it very useful and simple for collection initializers. that's the key point

**Reply by En3Tho** [reply]
@HaloFour I know C# won't.  It was just an example. F# has much terser syntax anyway and such things look quite better there (not to mention that my example used "yield" keyword which could be omitted).

The thing I like about inline iterators is that they can:
1) Be optimized better (e.g. if compiler can deduce that there are 5 elements added from the loop then it can inline that instead) 
2) Remain being structs and not get promoted to the heap in form of IEnumerable + IEnumerator or even better simply being inlined in the code right there

This is especially important now that .Net 8 has great struct promotion capabilities

I think overall this is a good addition to collection literals. I don't really like verbosity of it but at the same time IEnumerable syntax is the same so..

@CyrusNajmabadi What do you think about this?
